### ISSUE TITLE
v3: bug with parsing null values

### ISSUE THREAD

**Issue by rokkerruslan** [go-yaml/yaml#681]
Version 3 differs from version 2 in its behavior when parsing null (or the equivalent `empty/~`) values. Version 2 had behavior conforms to the specification (partially), in particular clause `7.2. Empty Nodes': 

 > YAML allows the node content to be omitted in many cases. Nodes with empty content are interpreted as if they were plain scalars with an empty value. Such nodes are commonly resolved to a "null" value. 

The `v3` version has incorrect behavior because it considers `null` values as missing keys. Example:

```go
package main

import (
	"fmt"

	v2 "gopkg.in/yaml.v2"
	v3 "gopkg.in/yaml.v3"
)

type Config struct {
	Fields map[string]string `yaml:"fields"`
}

const configYAML = `
fields:
        f1: null
        f2: 
`

func main() {
	cfg := Config{}
	err := v2.Unmarshal([]byte(configYAML), &cfg)
	if err != nil {
		panic(err)
	}
	fmt.Println("v2", cfg.Fields)

	cfg = Config{}
	err = v3.Unmarshal([]byte(configYAML), &cfg)
	if err != nil {
		panic(err)
	}
	fmt.Println("v3", cfg.Fields)
}

```

Expected:

```
v2 map[f1: f2:]
v3 map[f1: f2:]

// Or parsing error
```

Got:

```
v2 map[f1: f2:]
v3 map[]
```

Go playground with example - https://play.golang.org/p/Sf_o_7ErXVN

--------------------------------------------------

The json parser from the standard library has a similar behavior to `v2` - [Playground](https://play.golang.org/p/B1ynxYuez0T)

**Comment by niemeyer** [comment]
Thanks for the report. I'll have a look at this.

**Comment by niemeyer** [comment]
I'm hoping the change in c8046fb addresses the intended use cases. Let me explain the situation and what was done.

In v2, we had a pretty straightforward rule of resetting whatever value was in place before to its zero value, which creates the behavior described for v2 in the original report above: the map was empty, and there are two null fields, so both of them get set to the zero value of the map which is the empty string in this case. Very easy to understand, but has an issue: default values.

In v3, that original behavior was changed to preserve the state of the value being decoded into when the data doesn't match the particular value type. Since the map value is a string, a null is rejected and is simply skipped, as any other incorrect value type would be.

In encoding/json, the behavior is neither of these, and it's unfortunately not very consistent by itself either. The json module will leave a struct field value untouched if it's set. So, for example, an "N int" struct field that is set to 42 will remain at 42 when "n: null" is decoded into it. This matches the original v3 behavior above. But if that same scenario decodes into a map that already has an "N" field set to 42, that map will be mutated so "n" is set to the zero value of the map type, 0. This matches the behavior of v2 as described above.

So, because of that inconsistency, the behavior introduced by c8046fb is neither v2's nor the one in encoding/json. It's rather an attempt to introduce consistent behavior where null means leave the preset value whenever it cannot decode correctly into the value type, both in maps and structs. So when decoding "n: null" the behavior is consistent for maps and structs:

With struct fields:

1. Set to nil if that's a valid value for the type
2. Otherwise if the field is preset, leave that alone
3. If the field is missing, it will have the zero value already

With maps:

1. Set to nil if that's a valid value for the type
2. Otherwise if the key is preset, leave that alone
3. If the key is missing, set it to the zero value

I'm hoping this behavior will be clear and convenient in practice. Please let me know how you find it.

**Comment by dgl** [comment]
I feel the behaviour seen in v2 is more correct for YAML, I think trying to emulate the behaviour from encoding/json is actually a bit confusing -- YAML is not JSON and this very difference comes from the use of types.

In particular a null value in YAML can be kind of implicit, in that:

```yaml
foo:
```

Sets foo to null. Particularly when dealing with strings this means if I have `foo: bar` and then simply delete `bar`, one may expect that to be set to the empty string, as it was in v2.

In particular there are likely already configurations out there which rely on this, silently changing behaviour would essentially block upgrading to v3 I think.

I see some options here:
- Add a decoder option that makes nulls become the zero value, as v2 did;
- Add an option that errors on invalid type conversion, rather than simply ignoring it

**Comment by niemeyer** [comment]
That's indeed an interesting point, @dgl.  To be clear, I'm not worried about the change in behavior. There's a reason why we've been waiting for years to do a v3, instead of just updating v2. There is change in behavior. That said, the changes must be an improvement, so let's make sure this is the case here too.

So, let's see in which case this issue will be observed:

1. There must be a previous field, explicitly set with a default value
2. There must be yaml with a key named after said field, and without a value
3. It must be the intention of the user that the field remains with an empty string as a value

Okay, so the first point to note here is that point 1 is impossible in v2. You cannot have a default value preserved while using yaml.v2, period. It will always reset your entire value. That's already change in behavior, and one that makes it impossible to have a similar workflow to the one described above.

The second point is that while using v3, the developer must have explicitly set a default value for the scenario to have meaning, otherwise the value would still result in the empty string. In such cases, the usual reason why someone would leave the key without a value is precisely because they want that default value. In other words, having "foo:" and having nothing at all behave the same.

So, while I understand your concern, I'm failing to find actual use cases from experience where we'd see all of 1+2+3 being true, instead of being exactly the sort of use cases that would justify the behavior implemented.

Can you come up with some realistic examples that would show such a problem in practice?

**Comment by dgl** [comment]
The use case here is in Prometheus, I'm a bit confused about you saying point 1 is impossible as it uses that a lot, e.g.:
https://play.golang.org/p/bByZSx_QGOc

Running with v2 (just hit run) and it will overwrite the "default" value, change the import to v3 and it won't.

The actual place this is an issue is in a test that did `regex:` to get a value set to the empty string (https://github.com/prometheus/prometheus/pull/7142#discussion_r483934371), in that case there was some complexity involving UnmarshalYAML too, but the same thing applies.

**Comment by niemeyer** [comment]
> The use case here is in Prometheus, I'm a bit confused about you saying point 1 is impossible as it uses that a lot, e.g.:

Some examples:

https://play.golang.org/p/PbZIEG5D_pz

Then switch to v3 to see it working.

> The actual place this is an issue is in a test that did regex: to get a value set to the empty string

Can you explain the use case?  On a first uninformed sight, this looks quite awkward. I'd never guess that "regex:" line is playing a relevant role. As a user, my guess is precisely that this is enabling some sort of default behavior.

**Comment by niemeyer** [comment]
Sorry, the original example above missed the default string. I've fixed it and updated the link.

**Comment by dgl** [comment]
> Some examples [...]

Thanks, I see the difference, but I wouldn't say defaults aren't at all possible with v2 given that; it is just possible to structure a config in a way that people may easily clear parts...

Actually Prometheus manages to workaround that in a way that works on v2 and v3 (at least with #689... ;) ): https://github.com/prometheus/prometheus/blob/master/config/config.go#L176

> Can you explain the use case? On a first uninformed sight, this looks quite awkward. I'd never guess that "regex:" line is playing a relevant role. As a user, my guess is precisely that this is enabling some sort of default behavior.

The particular use case here is a bit odd as it's a test, but relabels in Prometheus are a way to run "actions" on input data, to do things like alter or drop it -- this could be used to drop data where a particular input field is empty (the regex is a full match, so the empty string means matching exactly the empty string, rather than any string as a plain regex would do).

I agree it's awkward and writing `regex: ""` is far clearer but Prometheus has a backwards compatibility promise, which this would break. (And potentially this could affect other fields, this one just happened to be found by a unit test.)

The map fix (thanks!) and the fix in #689 would actually solve the issues I know of (including the regex one) -- but the bit that worries me is this happens silently, so maybe having an option to error when a bad type conversion for a scalar is attempted would be the best approach.

**Comment by thernstig** [comment]
@niemeyer the Helm package uses go-yaml, but there seems to be a limit to combine this with the usage of creating nice-to-read Helm values.yaml files with good JSON schemas due to the parsing logic, see https://github.com/helm/helm/issues/11588

Is there any flag/solution fort his in go-yaml to fix that or would that be a valid feature request?
